Multi-server: adding server local placement in sched

agrawalravi90 opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->
In a multi-server setup, there can be a single scheduler for all server instances. Each node and each job is owned by a particular server. The scheduler gets a view of all nodes and all jobs and can potentially run a job anywhere. Running a job on nodes which don't belong to the same server incurs the cost of moving the job from one server to another. To avoid this, it's desirable to enhance scheduler to consider nodes of the server which owns a job before considering other nodes in the cluster, in the hope to minimize job moves between servers. This PR aims to do just that.

#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->
- Added a new resource called 'msvr_node_group' which is set by each server to the server id on all of its nodes
- In the absence of admin configured node grouping, for a multi-server setup, scheduler groups nodes by this msvr_node_group, which creates placement sets of nodes of each server, and caches it.
- When considering the nodes to use for a job, the scheduler looks up the cache of placement sets of nodes for each server and creates a pset array with that pset as the first element. If the job is not an array, and is single chunk, it also adds the allpart as the second element of this array, and then passes it to eval_selspec, which means that the pset of the local server gets picked first before sched tries non local nodes.
- Right now, we are going to restrict job arrays and multi-chunk jobs to not span across servers

#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->
the following test shows how jobs are first scheduled on their owner's nodes and are only placed on nodes of another server in the absence of resources on the owner server
```
[ravi@pbsc8 ~]$ pbsnodes -av
pbsc8
     Mom = pbsc8
     ntype = PBS
     state = free
     pcpus = 4
     resources_available.arch = linux
     resources_available.host = pbsc8
     resources_available.mem = 2035188kb
     resources_available.msvr_node_group = pbsc8
     resources_available.ncpus = 4
     resources_available.vnode = pbsc8
     resources_assigned.accelerator_memory = 0kb
     resources_assigned.hbmem = 0kb
     resources_assigned.mem = 0kb
     resources_assigned.naccelerators = 0
     resources_assigned.ncpus = 0
     resources_assigned.vmem = 0kb
     resv_enable = True
     sharing = default_shared
     license = l
     last_state_change_time = Thu Dec 10 00:22:05 2020
     last_used_time = Thu Dec 10 00:22:11 2020

pbsc8-2
     Mom = pbsc8
     ntype = PBS
     state = free
     pcpus = 4
     jobs = 4.pbsc8-2/0, 5.pbsc8-2/1
     resources_available.arch = linux
     resources_available.host = pbsc8
     resources_available.mem = 2035188kb
     resources_available.msvr_node_group = pbsc8-2
     resources_available.ncpus = 4
     resources_available.vnode = pbsc8-2
     resources_assigned.accelerator_memory = 0kb
     resources_assigned.hbmem = 0kb
     resources_assigned.mem = 0kb
     resources_assigned.naccelerators = 0
     resources_assigned.ncpus = 2
     resources_assigned.vmem = 0kb
     resv_enable = True
     sharing = default_shared
     license = l
     last_state_change_time = Thu Dec 10 00:22:05 2020
     last_used_time = Thu Dec 10 00:22:05 2020

[ravi@pbsc8 ~]$ qsub -- /bin/sleep 100
6.pbsc8-2
[ravi@pbsc8 ~]$ qsub -- /bin/sleep 100
7.pbsc8-2
[ravi@pbsc8 ~]$ qstat
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
4.pbsc8-2         STDIN            ravi              00:00:00 R workq           
5.pbsc8-2         STDIN            ravi              00:00:00 R workq           
6.pbsc8-2         STDIN            ravi              00:00:00 R workq           
7.pbsc8-2         STDIN            ravi              00:00:00 R workq           
[ravi@pbsc8 ~]$ pbsnodes -av
pbsc8
     Mom = pbsc8
     ntype = PBS
     state = free
     pcpus = 4
     resources_available.arch = linux
     resources_available.host = pbsc8
     resources_available.mem = 2035188kb
     resources_available.msvr_node_group = pbsc8
     resources_available.ncpus = 4
     resources_available.vnode = pbsc8
     resources_assigned.accelerator_memory = 0kb
     resources_assigned.hbmem = 0kb
     resources_assigned.mem = 0kb
     resources_assigned.naccelerators = 0
     resources_assigned.ncpus = 0
     resources_assigned.vmem = 0kb
     resv_enable = True
     sharing = default_shared
     license = l
     last_state_change_time = Thu Dec 10 00:22:05 2020
     last_used_time = Thu Dec 10 00:22:11 2020

pbsc8-2
     Mom = pbsc8
     ntype = PBS
     state = job-busy
     pcpus = 4
     jobs = 4.pbsc8-2/0, 5.pbsc8-2/1, 6.pbsc8-2/2, 7.pbsc8-2/3
     resources_available.arch = linux
     resources_available.host = pbsc8
     resources_available.mem = 2035188kb
     resources_available.msvr_node_group = pbsc8-2
     resources_available.ncpus = 4
     resources_available.vnode = pbsc8-2
     resources_assigned.accelerator_memory = 0kb
     resources_assigned.hbmem = 0kb
     resources_assigned.mem = 0kb
     resources_assigned.naccelerators = 0
     resources_assigned.ncpus = 4
     resources_assigned.vmem = 0kb
     resv_enable = True
     sharing = default_shared
     license = l
     last_state_change_time = Thu Dec 10 00:22:57 2020
     last_used_time = Thu Dec 10 00:22:05 2020

[ravi@pbsc8 ~]$ qsub -- /bin/sleep 100
8.pbsc8-2
[ravi@pbsc8 ~]$ qsub -- /bin/sleep 100
9.pbsc8-2
[ravi@pbsc8 ~]$ pbsnodes -av
pbsc8
     Mom = pbsc8
     ntype = PBS
     state = free
     pcpus = 4
     jobs = 8.pbsc8-2/0, 9.pbsc8-2/1
     resources_available.arch = linux
     resources_available.host = pbsc8
     resources_available.mem = 2035188kb
     resources_available.msvr_node_group = pbsc8
     resources_available.ncpus = 4
     resources_available.vnode = pbsc8
     resources_assigned.accelerator_memory = 0kb
     resources_assigned.hbmem = 0kb
     resources_assigned.mem = 0kb
     resources_assigned.naccelerators = 0
     resources_assigned.ncpus = 2
     resources_assigned.vmem = 0kb
     resv_enable = True
     sharing = default_shared
     license = l
     last_state_change_time = Thu Dec 10 00:22:05 2020
     last_used_time = Thu Dec 10 00:22:11 2020

pbsc8-2
     Mom = pbsc8
     ntype = PBS
     state = job-busy
     pcpus = 4
     jobs = 4.pbsc8-2/0, 5.pbsc8-2/1, 6.pbsc8-2/2, 7.pbsc8-2/3
     resources_available.arch = linux
     resources_available.host = pbsc8
     resources_available.mem = 2035188kb
     resources_available.msvr_node_group = pbsc8-2
     resources_available.ncpus = 4
     resources_available.vnode = pbsc8-2
     resources_assigned.accelerator_memory = 0kb
     resources_assigned.hbmem = 0kb
     resources_assigned.mem = 0kb
     resources_assigned.naccelerators = 0
     resources_assigned.ncpus = 4
     resources_assigned.vmem = 0kb
     resv_enable = True
     sharing = default_shared
     license = l
     last_state_change_time = Thu Dec 10 00:22:57 2020
     last_used_time = Thu Dec 10 00:22:05 2020
```


<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

agrawalravi90 commented:
@bhroam thanks for the feedback offline, I've changed the approach to  better use the placement set infra, please let me know how it looks.

arungrover commented on src/scheduler/job_info.cpp:
```diff
@@ -3116,7 +3057,7 @@ find_and_preempt_jobs(status *policy, int pbs_sd, resource_resv *hjob, server_in
 		}
 
-		if ((preempt_jobs_reply = pbs_preempt_jobs(pbs_sd, preempt_jobs_list)) == NULL) {
+		if ((preempt_jobs_reply = send_preempt_jobs(pbs_sd, preempt_jobs_list)) == NULL) {
```
We should also add a new type of sort to make sure that list of running jobs list has local jobs in the beginning of the list and jobs belonging to other servers behind. We already sort jobs on the basis of min_time_since_start. Maybe we will have to do a multi-sort of thing where you first sort jobs on the basis of server and when the server is same sort on their start time.

agrawalravi90 replied:
so that we preempt jobs from the local server first? Ya I think that will be useful, I hadn't thought about that, thanks for mentioning it. Is it ok if we take that up separately?

arungrover replied:
Fine with me, Can you please log a ticket?

agrawalravi90 commented:
@ashwathraop  and @arungrover please provide further feedback, thanks!

arungrover commented on src/scheduler/check.cpp:
```diff
@@ -1566,13 +1570,36 @@ check_normal_node_path(status *policy, server_info *sinfo, queue_info *qinfo, re
+				for (i = 0; sinfo->svr_to_psets[i] != NULL; i++) {
+					if (strcmp(sinfo->svr_to_psets[i]->svr_inst_id, resresv->job->svr_inst_id) == 0) {
+						msvr_pset[0] = sinfo->svr_to_psets[i]->np;
+						if (!resresv->job->is_subjob && spec->total_chunks == 1) {
```
I think is_subjob won't be true here, because subjobs are created in run_update_resresv. So you might want to check for is_array parent instead.

agrawalravi90 replied:
thanks for catching this and suggesting an alternative

arungrover commented on src/scheduler/fifo.cpp:
```diff
@@ -1356,7 +1327,8 @@ send_run_job(int pbs_sd, int has_runjob_hook, char *jobid, char *execvnode, char
  */
 int
-run_job(int pbs_sd, resource_resv *rjob, char *execvnode, int has_runjob_hook, char *node_owner, schd_error *err)
+run_job(int pbs_sd, resource_resv *rjob, char *execvnode, int has_runjob_hook, schd_error *err,
```
Please update the function header

arungrover commented on src/scheduler/node_partition.cpp:
```diff
@@ -1403,3 +1405,60 @@ update_all_nodepart(status *policy, server_info *sinfo, unsigned int flags)
+/**
+ * @brief	Constructor for a server_psets array
+ *
+ * @param	nothing
```
Please update this header.

arungrover commented on src/scheduler/node_partition.cpp:
```diff
@@ -1403,3 +1405,60 @@ update_all_nodepart(status *policy, server_info *sinfo, unsigned int flags)
 }
+
+/**
+ * @brief	Constructor for a server_psets array
```
This doesn't look like a copy constructor.

arungrover reviewed (changes requested):
Minor comments

ashwathraop commented on src/scheduler/server_info.cpp:
```diff
@@ -460,6 +461,34 @@ query_server(status *pol, int pbs_sd)
 	 */
 	create_placement_sets(policy, sinfo);
+	if (!sinfo->node_group_enable && sinfo->node_group_key != NULL &&
+			strcmp(sinfo->node_group_key[0], "msvr_node_group") == 0) {
```
This might be a dumb question, "msvr_node_group" sounds very generic to me. What happens if the admin sets the same name to one of his placement sets?

Also, you might want to make it a #define or a const variable since you used it in more than one place.

agrawalravi90 replied:
Hmm .. admins set pset to resource names that they can set on the nodes .. msvr_node_group is a resource .. so they won’t be able to create another resource with the same name .. msvr_node_group is also not settable, so even if they read about it in the guides and tried to use it, they won’t be able to .. so i think it should be okay .. let me know what you think

I should be able to use 'RESC_MSVR_ND_GROUP' instead of the string literal, I'll change it thanks

agrawalravi90 commented on src/scheduler/buckets.cpp:
```diff
@@ -1132,11 +1132,24 @@ check_node_buckets(status *policy, server_info *sinfo, queue_info *qinfo, resour
+			if (strcmp(sinfo->svr_to_psets[i]->svr_inst_id, resresv->job->svr_inst_id) == 0) {
+				nspec **nspecs;
+
+				nspecs = map_buckets(policy, sinfo->svr_to_psets[i]->np->bkts, resresv, err);
```
this might have been deliberate to not confuse server_info's buckets with a node partition's .. I didn't write it originally, so I'd prefer leaving it as is unless you feel strongly about it.

bhroam commented on src/scheduler/node_partition.cpp:
```diff
@@ -1403,3 +1405,60 @@ update_all_nodepart(status *policy, server_info *sinfo, unsigned int flags)
+	server_psets **ret = NULL;
+	int i;
+
+	ret = static_cast<server_psets **> (malloc((num_psets + 1) * sizeof(server_psets *)));
```
We're C++ now, and this is completely new.  Consider using a vector.

agrawalravi90 replied:
ok ill give it a shot

bhroam commented on src/scheduler/sched_ifl_wrappers.cpp:
```diff
@@ -0,0 +1,156 @@
+ * @param[in]	has_runjob_hook	- does server have a runjob hook?
+ * @param[in]	jobid	-	id of the job to run
+ * @param[in]	execvnode	-	the execvnode to run the job on
+ * @param[in]	node_owner	-	node owning server of first node in execvnode
```
Rename this

bhroam commented on src/scheduler/sched_ifl_wrappers.cpp:
```diff
@@ -0,0 +1,156 @@
+ */
+int
+send_run_job(int virtual_sd, int has_runjob_hook, char *jobid, char *execvnode,
+ 	     char *svr_id_node, char *svr_id_job)
```
Add the new param to the doxygen header

bhroam commented on src/scheduler/sched_ifl_wrappers.cpp:
```diff
@@ -0,0 +1,156 @@
+  	job_owner_sd = get_svr_inst_fd(virtual_sd, svr_id_job);
+
+  	extend[0] = '\0';
+ 	if (svr_id_node && svr_id_job && strcmp(svr_id_node, svr_id_job) != 0)
```
Check pointers against NULL

bhroam commented on src/scheduler/sched_ifl_wrappers.cpp:
```diff
@@ -0,0 +1,156 @@
+send_preempt_jobs(int virtual_sd, char **preempt_jobs_list)
+{
+    return pbs_preempt_jobs(virtual_sd, preempt_jobs_list);
+}
```
Why do we need this function?

agrawalravi90 replied:
I wanted to create IFL wrappers for the scheduler so that all of them could be found in one place. I had a hard time ensuring that all IFL calls from scheduler were correct as they were scattered throughout the code, so I thought that this will be useful. Let me know what you think.

bhroam commented on src/scheduler/server_info.cpp:
```diff
@@ -644,6 +673,15 @@ query_server_info(status *pol, struct batch_status *server)
 
+	if (sinfo->node_group_key == NULL and get_num_servers() > 1) {
+		/* Set node_group_key to msvr_node_group for server local placement */
+		sinfo->node_group_key = break_comma_list((char *) "msvr_node_group");
```
This should be const_cast<char *>("mscr_node_group")

bhroam commented on src/scheduler/node_partition.cpp:
```diff
@@ -1403,3 +1405,60 @@ update_all_nodepart(status *policy, server_info *sinfo, unsigned int flags)
+ * @retval	NULL for malloc error
+ */
+server_psets **
+new_server_psets(int num_psets)
```
Turn this into a true constructor

agrawalravi90 replied:
not needed anymore

bhroam commented on src/scheduler/node_partition.cpp:
```diff
@@ -1403,3 +1405,60 @@ update_all_nodepart(status *policy, server_info *sinfo, unsigned int flags)
+ * @return	void
+ */
+void
+free_server_psets(server_psets **arr)
```
Turn this into a true destructor

agrawalravi90 replied:
not needed anymore

bhroam commented on src/scheduler/node_partition.cpp:
```diff
@@ -1403,3 +1405,60 @@ update_all_nodepart(status *policy, server_info *sinfo, unsigned int flags)
+		}
+		free(arr);
+	}
+}
```
You don't have a copy constructor

agrawalravi90 replied:
not needed anymore

agrawalravi90 commented:
@arungrover  @bhroam  @ashwathraop  I think I've addressed all of your comments, please let me know if i missed any. Please provide further feedback, thanks

agrawalravi90 commented:
Sorry i had to rebase, i somehow had commits from my preemption pr

arungrover commented on src/scheduler/check.cpp:
```diff
@@ -1570,13 +1570,12 @@ check_normal_node_path(status *policy, server_info *sinfo, queue_info *qinfo, re
-				for (i = 0; sinfo->svr_to_psets[i] != NULL; i++) {
-					if (strcmp(sinfo->svr_to_psets[i]->svr_inst_id, resresv->job->svr_inst_id) == 0) {
-						msvr_pset[0] = sinfo->svr_to_psets[i]->np;
+				for (auto spset = sinfo->svr_to_psets.begin();
```
you can simplify it to just say "for (auto spset: sinfo->svr_to_psets)" instead.

vstumpf commented on src/scheduler/buckets.cpp:
```diff
@@ -1134,11 +1134,11 @@ check_node_buckets(status *policy, server_info *sinfo, queue_info *qinfo, resour
 		/* Find buckets associated with nodes of the server which owns the job */
-		for (auto spset = sinfo->svr_to_psets.begin(); spset != sinfo->svr_to_psets.end(); spset++) {
-			if (strcmp(spset->svr_inst_id, resresv->job->svr_inst_id) == 0) {
+		for (auto spset: sinfo->svr_to_psets) {
```
this makes a copy of each pset in `sinfo->svr_to_psets`.
If you don't want to do this, you can iterate over references
`for (auto &spset : sinfo->svr_to_psets) {`

See [StackOverflow](https://stackoverflow.com/questions/15176104/c11-range-based-loop-get-item-by-value-or-reference-to-const) for a better explanation

agrawalravi90 replied:
thanks for educating! will change it to use references

bhroam commented on src/lib/Libattr/master_resc_def_all.xml:
```diff
@@ -951,6 +951,20 @@
+      <member_at_comp>comp_arst</member_at_comp>
+      <member_at_free>free_arst</member_at_free>
+      <member_at_action>NULL_FUNC_RESC</member_at_action>
+      <member_at_flags>ATR_DFLAG_SvRD | ATR_DFLAG_SvWR | ATR_DFLAG_CVTSLT</member_at_flags>
```
By making this SvRD and SvWR, no one bug the server and scheduler can read it.  This means pbs_est will not have access.

agrawalravi90 replied:
if pbs_est is run with the sched user, then it should right?

bhroam replied:
It's not the user which determines whether or not these attributes are sent.  It has to do with the fact that it is a PBS daemon.  Until we allow pbs_est to register as a scheduler, it won't see any SvWR attributes.

bhroam reviewed (approved):
LGTM.  Thanks for using a vector!

bhroam commented:
When you get a chance, can you resolve the conflicts?

vstumpf commented on src/scheduler/server_info.cpp:
```diff
@@ -644,6 +667,15 @@ query_server_info(status *pol, struct batch_status *server)
 	site_set_share_head(sinfo);
 #endif /* localmod 034 */
 
+	if (sinfo->node_group_key == NULL and get_num_servers() > 1) {
```
I think this needs to be &&

bhroam replied:
Yeah, it'll need to copy it, otherwise calendaring will all get messed up.

vstumpf reviewed (changes requested):
does `dup_server_info()` need to copy the `svr_to_psets` vector?

agrawalravi90 commented:
@vstumpf  thanks for catching important issues, ive addressed them, please take a look

@bhroam  let me know if the last commit looks good, thanks

vstumpf reviewed (approved):
LGTM
